pkg/cloudsources/ocm/client.go

```diff
@@ -0,0 +1,137 @@
+package ocm
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	gogoProto "github.com/gogo/protobuf/types"
+	sdkClient "github.com/openshift-online/ocm-sdk-go"
+	accountsmgmtv1 "github.com/openshift-online/ocm-sdk-go/accountsmgmt/v1"
+	pkgErrors "github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/clientconn"
+	"github.com/stackrox/rox/pkg/cloudsources/discoveredclusters"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/urlfmt"
+)
+
+type ocmClient struct {
+	conn          *sdkClient.Connection
+	cloudSourceID string
+}
+
+// NewClient creates a client to interact with OCM APIs.
+func NewClient(config *storage.CloudSource) (*ocmClient, error) {
+	connection, err := sdkClient.NewConnectionBuilder().
+		URL(urlfmt.FormatURL(config.GetOcm().GetEndpoint(), urlfmt.HTTPS, urlfmt.NoTrailingSlash)).
+		Tokens(config.GetCredentials().GetSecret()).Agent(clientconn.GetUserAgent()).Build()
+
+	if err != nil {
+		return nil, pkgErrors.Wrap(err, "creating OCM connection")
+	}
+
+	return &ocmClient{
+		conn:          connection,
+		cloudSourceID: config.GetId(),
+	}, nil
+}
+
+func (c *ocmClient) GetDiscoveredClusters(ctx context.Context) ([]*discoveredclusters.DiscoveredCluster, error) {
+	var (
+		subs  []*accountsmgmtv1.Subscription
+		total int
+		page  = 1
+	)
+
+	// Taken from console.redhat.com/openshift.
+	// Filter for all subscriptions which have:
+	//	- a cluster associated with it
+	// 	- the cluster is a valid OpenShift plan
+	//	- the status is "Active", "Disconnected"
+	//  - name / cluster_id / external_cluster_id are given
+	subscriptionSearch := "(cluster_id!='') " +
+		"AND (plan.id IN ('ARO', 'OCP', 'MOA', 'OCP-AssistedInstall', 'MOA-HostedControlPlane', 'OSD', 'OSDTrial')) " +
+		"AND (status IN  ('Active', 'Disconnected')) " +
+		"AND (display_name ILIKE '%%' OR external_cluster_id ILIKE '%%' OR cluster_id ILIKE '%%')"
+
+	for {
+		resp, err := c.conn.AccountsMgmt().V1().Subscriptions().List().Size(100).Page(page).Search(subscriptionSearch).SendContext(ctx)
```

Comment:
Can you add a comment here why we are not using `/api/clusters_mgmt/v1/clusters`?